server.go

```diff
@@ -465,12 +516,17 @@ func (srv *Server) serveTCP(l net.Listener) error {
 			}
 			return err
 		}
-		go srv.serveTCPConn(handler, rw)
+		w := &response{tsigSecret: srv.TsigSecret, tcp: rw}
+		select {
+		case srv.queueTCPConn <- w:
+		default:
+			srv.spawnWorkerTCP()
+			srv.queueTCPConn <- w
```

Comment:
There’s a race condition here. If another request comes in and reaches the `select` statement above *before* this line is reached, it may steal the new worker.

A solution to this is simple, pass `w` directly to `spawnWorkerTCP`. If the argument is non-nil, it can be handled before the `for range` loop.

Reply:
Don't follow about race - this is one thread and srv.queue is unbuffered channel,
so we cannot reach select until we push job to the queue.

Reply:
@UladzimirTrehubenka It's a race with a *separate* request that hits the `select`. The spawning and sending are not atomic operations as a set.

This isn't a correctness issue, but it could lead to hard to diagnose latency spikes.

Reply:
@UladzimirTrehubenka You're dead right actually, ignore me. I didn't realise this was only ever called from one goroutine.

You could still see, an admittedly small, performance improvement by skipping the channel send as I suggested.

Reply:
Done